python/ray/actor.py

```diff
@@ -802,97 +837,55 @@ def __ray_restore_from_checkpoint__(cls, pickled_checkpoint):
-                        worker.actors[worker.actor_id] = actor
-                    # An exception was thrown. Save the error.
-                    except Exception as error:
-                        # We could not resume the checkpoint, so count the task
-                        # as failed.
-                        actor_checkpoint_failed = True
-                        error_to_return = error
-                else:
-                    # We cannot resume a mismatching checkpoint, so count the
-                    # task as failed.
-                    actor_checkpoint_failed = True
-
-            # Fall back to lineage reconstruction if we were unable to load the
-            # checkpoint.
-            if actor_checkpoint_failed:
-                worker.local_scheduler_client.reconstruct_object(
-                    plasma_id.binary())
-                worker.local_scheduler_client.notify_unblocked()
-
-            return actor_checkpoint_failed, error_to_return
+            checkpoint_index = worker.actor_task_counter
+            # Get the state to save.
+            checkpoint = self.__ray_save_checkpoint__()
+            # Get the current task frontier, per actor handle.
+            # NOTE(swang): This only includes actor handles that the local
+            # scheduler has seen. Handle IDs for which no task has yet reached
+            # the local scheduler will not be included, and may not be runnable
+            # on checkpoint resumption.
+            actor_id = ray.local_scheduler.ObjectID(worker.actor_id)
+            frontier = worker.local_scheduler_client.get_actor_frontier(
+                actor_id)
+            # Save the checkpoint in Redis. TODO(rkn): Checkpoints
+            # should not be stored in Redis. Fix this.
+            set_actor_checkpoint(worker, worker.actor_id, checkpoint_index,
+                                 checkpoint, frontier)
+
+        def __ray_checkpoint_restore__(self):
+            """Restore a checkpoint.
+
+            This task looks for a saved checkpoint and if found, restores the
+            state of the actor, the task frontier in the local scheduler, and
+            the checkpoint index (number of tasks executed so far).
+            """
+            worker = ray.worker.global_worker
+            # Get the most recent checkpoint stored, if any.
+            checkpoint_index, checkpoint, frontier = get_actor_checkpoint(
+                worker, worker.actor_id)
+            # Try to resume from the checkpoint.
+            checkpoint_resumed = False
+            if checkpoint_index is not None:
+                # Load the actor state from the checkpoint.
+                worker.actors[worker.actor_id] = (
+                    worker.actor_class.__ray_restore_from_checkpoint__(
+                        checkpoint))
+                # Set the number of tasks executed so far.
+                worker.actor_task_counter = int(checkpoint_index)
```

Comment:
Instead of casting to `int` here, this cast should probably be done inside of `get_actor_checkpoint`